pkg/credentialprovider/plugin/plugin.go

```diff
@@ -193,12 +199,27 @@ func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
 		return cachedConfig
 	}
 
-	response, err := p.plugin.ExecPlugin(context.Background(), image)
+	// ExecPlugin is wrapped in single flight to exec plugin once for concurrent same image request.
+	// The caveat here is we don't know cacheKeyType yet, so if cacheKeyType is registry/global and credentials saved in cache
+	// on per registry/global basis then exec will be called for all requests if requests are made concurrently.
+	// foo.bar.registry
+	// foo.bar.registry/image1
+	// foo.bar.registry/image2
+	res, err, _ := p.group.Do(image, func() (interface{}, error) {
+		return p.plugin.ExecPlugin(context.Background(), image)
```

Comment:
This does assume that a plugin can handle concurrent execution, i.e. it uses some file based locking to prevent itself from concurrently writing to a local cred file or similar.  Do the docs make that clear?

Reply:
sorry I might not have understood here but Kubelet is managing the parallel execution here, creating new buffers for each request https://github.com/kubernetes/kubernetes/blob/b96c86f1b9d92880845724d2030da3e2adac89e5/pkg/credentialprovider/plugin/plugin.go#L334-L353

Reply:
@adisky I am referring to parallel execution of the exec plugin itself, meaning that the binary may be run multiple times at the same time.  I think that is already the case even without this change.

Reply:
Actually I think the locking prevented parallel execution previously.  So I do think this behavior should be documented, if it is not already.

Reply:
Ack, will update the documentation.